src/backend/distributed/utils/shardinterval_utils.c

```diff
@@ -347,10 +347,22 @@ FindShardIntervalIndex(Datum searchedValue, CitusTableCacheEntry *cacheEntry)
 		}
 		else
 		{
-			int hashedValue = DatumGetInt32(searchedValue);
-			uint64 hashTokenIncrement = HASH_TOKEN_COUNT / shardCount;
+			/*
+			 * We use a 64-bit integer to be able to perform arithmetic
+			 * without risk of overflow.
+			 */
+			int64 hashedValue = DatumGetInt32(searchedValue);
+
+			/* normalize to the 0-UINT32_MAX range */
+			int64 normalizedHashValue = hashedValue - INT32_MIN;
+
+			/* size of each hash range */
+			int64 hashRangeSize = HASH_TOKEN_COUNT / shardCount;
+
+			/* index of hash range into which the hash value falls */
+			shardIndex = (uint32) (normalizedHashValue / hashRangeSize);
 
-			shardIndex = (uint32) (hashedValue - INT32_MIN) / hashTokenIncrement;
+			Assert(shardIndex >= 0);
```

Comment:
is the type of shardIndex `int` instead of `uint32` to have `INVALID_SHARD_INDEX  = -1` ?

Reply:
yes